internal/dataplane/kongstate/kongstate.go

```diff
@@ -330,81 +334,147 @@ func (ks *KongState) FillVaults(
-		pluginKey := namespace + ":" + pluginName
-		relations, ok := pluginRels[pluginKey]
-		if !ok {
-			relations = util.ForeignRelations{}
-		}
-		relations.Service = append(relations.Service, identifier)
 		pluginRels[pluginKey] = relations
 	}
 
 	for i := range ks.Services {
-		// service
 		for _, svc := range ks.Services[i].K8sServices {
-			pluginList := annotations.ExtractKongPluginsFromAnnotations(svc.GetAnnotations())
-			for _, pluginName := range pluginList {
-				addServiceRelation(svc.Namespace, pluginName, *ks.Services[i].Name)
+			pluginList := annotations.ExtractNamespacedKongPluginsFromAnnotations(svc.GetAnnotations())
+			for _, plugin := range pluginList {
+				addRelation(svc, plugin, *ks.Services[i].Name, ServiceRelation)
 			}
 		}
-		// route
+
 		for j := range ks.Services[i].Routes {
 			ingress := ks.Services[i].Routes[j].Ingress
-			pluginList := annotations.ExtractKongPluginsFromAnnotations(ingress.Annotations)
-			for _, pluginName := range pluginList {
-				addRouteRelation(ingress.Namespace, pluginName, *ks.Services[i].Routes[j].Name)
+			pluginList := annotations.ExtractNamespacedKongPluginsFromAnnotations(ingress.Annotations)
+			for _, plugin := range pluginList {
+				// pretend we have a full Ingress struct for reference checks
+				virtualIngress := netv1.Ingress{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: ingress.Namespace,
+						Name:      ingress.Name,
+					},
+				}
+				addRelation(&virtualIngress, plugin, *ks.Services[i].Routes[j].Name, RouteRelation)
 			}
 		}
 	}
-	// consumer
+
 	for _, c := range ks.Consumers {
-		pluginList := annotations.ExtractKongPluginsFromAnnotations(c.K8sKongConsumer.GetAnnotations())
-		for _, pluginName := range pluginList {
-			addConsumerRelation(c.K8sKongConsumer.Namespace, pluginName, *c.Username)
+		pluginList := annotations.ExtractNamespacedKongPluginsFromAnnotations(c.K8sKongConsumer.GetAnnotations())
+		for _, plugin := range pluginList {
+			addRelation(&c.K8sKongConsumer, plugin, *c.Username, ConsumerRelation)
 		}
 	}
-	// consumer group
+
 	for _, cg := range ks.ConsumerGroups {
-		pluginList := annotations.ExtractKongPluginsFromAnnotations(cg.K8sKongConsumerGroup.GetAnnotations())
-		for _, pluginName := range pluginList {
-			addConsumerGroupRelation(cg.K8sKongConsumerGroup.Namespace, pluginName, *cg.Name)
+		pluginList := annotations.ExtractNamespacedKongPluginsFromAnnotations(cg.K8sKongConsumerGroup.GetAnnotations())
+		for _, plugin := range pluginList {
+			addRelation(&cg.K8sKongConsumerGroup, plugin, *cg.Name, ConsumerGroupRelation)
 		}
 	}
 
 	return pluginRels
 }
 
+type pluginReference struct {
+	Referer   client.Object
+	Namespace string
+	Name      string
+}
+
+func isRemotePluginReferenceAllowed(s store.Storer, r pluginReference) error {
+	// remote plugin. considered part of this namespace if a suitable ReferenceGrant exists
+	grants, err := s.ListReferenceGrants()
```

Comment:
Are there any performance concerns to run `ListReferenceGrants()` each time when a resource (`Service`/`Consumer`/...) refers to remote plugins? If we believe that there will not be significant amount of cross-namespace reference, this may be OK.

Reply:
I think it's mostly an upstream concern--we don't do much other than piggyback off the https://pkg.go.dev/k8s.io/client-go/tools/cache#Store `List()` implementations, and basically trust that the Kubernetes core devs make those calls efficient because they're used for everything, and would provide additional filters or special cases if it became a problem (similar to https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#motivation).

On our end all we can really do is ensure that we're not checking grants if we don't need to--we wouldn't want to call `ListReferenceGrants()` for every plugin, but if it's a remote plugin, we know we need to check them.